### PR TITLE
Package update: dont upload unmodified files aka dedupe

### PR DESCRIPTION
**Pull request by nl0**

### CODE REVIEW

**Review by fiskus — APPROVED** [review]
I checked the feature in the browser, didn't spot any bug. There is a control "Undo changes ←" when all files are the same (hashed), but I think, it's ok.

Code is good, just advice: I think that the better strategy for typescript migration is converting small modules to typescript instead of large ones. If you don't have small modules, you can create them by splitting large into small ones. Also, as you are using `.js` imports you can create intermediate types (`PackageDialog/types.ts`). Otherwise, you are forced to use `any` and typecasting.
In short: having a small number of tiny modules completely converted to typescript is better, than large ones converted partially with `any` and typecasting.

File: catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx [hunk 1]
```diff
@@ -129,7 +171,14 @@ const insertIntoDir = (path: string[], file: FilesEntry, dir: FilesEntryDir) =>
   const newChildren = insertIntoTree(path, file, children)
   const type = newChildren
     .map(FilesEntry.match({ Dir: R.prop('type'), File: R.prop('type') }))
-    .reduce((acc, entryType) => (acc === entryType ? acc : 'modified'))
+    .reduce((acc, entryType) =>
+      // eslint-disable-next-line no-nested-ternary
```
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:175]: this rule is a part of the config we extend and i didnt want to add another override, but i think using ternaries deliberately makes sense. in this case i'll cave in to it and use ifs, tho i think it's a little more elegant with ternaries, because it is an expression and you dont have to use returns (tbh i think that's one of the most annoying part of js/ts which coffescript, where almost anything is an expression, got right)

File: catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx [hunk 2]
```diff
@@ -167,19 +216,28 @@ interface IntermediateEntry {
...
-    if (isDir) {
-      return { type: 'unchanged' as const, path, size, isDir }
-    }
-    return { type: 'unchanged' as const, path, size }
-  })
+  // TODO: use hash.{value,ready} props to descriminate unchanged | added | hashing
+  const existingEntries = Object.entries(existing).map(
+    ([path, { isDir, size, hash }]) => {
+      if (path in deleted) {
+        return { type: 'deleted' as const, path, size }
```
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:223]: why? const strings make more sense imo, and you can also use them as keys (like, for classes or some matches), and it's properly type-checked anyways

File: catalog/app/containers/Bucket/PackageDialog/PackageDialog.tsx [hunk 3]
```diff
@@ -31,41 +32,42 @@ export const ERROR_MESSAGES = {
...
-    // eslint-disable-next-line consistent-return
-    return Array.from(new Uint8Array(hashBuf))
-      .map((b) => b.toString(16).padStart(2, '0'))
-      .join('')
-  } catch (e) {
-    // return undefined on error
-  }
+export const getNormalizedPath = (f: { path?: string; name: string }) => {
+  const p = f.path || f.name
+  return p.startsWith('/') ? R.drop(1, p) : p
```
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Bucket/PackageDialog/PackageDialog.tsx:37]: well, i dont think that's better, that's just more imperative and verbose, which is quite opposite of "better" in my book, but i agree that `.substring()` is more appropriate here

File: catalog/app/containers/Bucket/PackageDialog/PackageDialog.tsx [hunk 4]
```diff
@@ -514,7 +593,23 @@ export const MetaInput = React.forwardRef(function MetaInput(
...
-export function SchemaFetcher({ manifest, workflow, workflowsConfig, children }) {
+interface SchemaFetcherProps {
+  manifest?: {
+    workflow?: {
+      id?: string
+    }
+  }
+  workflow?: workflows.Workflow
+  workflowsConfig: workflows.WorkflowsConfig
+  children: (result: any) => React.ReactElement
```
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Bucket/PackageDialog/PackageDialog.tsx:604]: ok, but it wont help much. the actual thing required here is making AsyncResult and Data type-safe

File: catalog/app/containers/Bucket/PackageUpdateDialog.tsx [hunk 5]
```diff
@@ -18,15 +18,17 @@ import pipeThru from 'utils/pipeThru'
...
 
 import * as PD from './PackageDialog'
 import * as requests from './requests'
 
-type WorkflowsConfig = $TSFixMe
-
 interface Manifest {
   entries: Record<string, PD.ExistingFile>
-  meta: $TSFixMe
+  meta: {}
```
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Bucket/PackageUpdateDialog.tsx:28]: yep, i think we need to properly type all "core" quilt entities like packages and manifests and stuff and build everything else upon these foundational types (and converting to/from any alternative/domain-specific representations when necessary)

File: catalog/app/containers/Bucket/PackageUpdateDialog.tsx [hunk 6]
```diff
@@ -216,25 +223,29 @@ function DialogForm({
...
+      R.zipWith<typeof toUpload[number], UploadResult, Zipped>((f, r) => [
+        f.path,
+        {
+          physicalKey: s3paths.handleToS3Url({
+            bucket,
+            key: r.Key,
+            version: r.VersionId,
+          }),
+          size: f.file.size,
+          hash: f.file.hash.value!,
```
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Bucket/PackageUpdateDialog.tsx:240]: ok, i've added an invariant check for this

File: catalog/app/containers/Bucket/PackageUpdateDialog.tsx [hunk 7]
```diff
@@ -600,11 +617,14 @@ const DialogState = tagged.create(
...
     Loading: () => {},
     Error: (e: Error) => e,
-    Form: (v: { manifest: Manifest; workflowsConfig: WorkflowsConfig }) => v,
+    Form: (v: { manifest: Manifest; workflowsConfig: workflows.WorkflowsConfig }) => v,
     Success: (v: { name: string; hash: string }) => v,
   },
 )
 
+// eslint-disable-next-line @typescript-eslint/no-redeclare
+type DialogState = tagged.InstanceOf<typeof DialogState>
```
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Bucket/PackageUpdateDialog.tsx:626]: first, in this case it's getting imported together with the value, so no need for another import;
second, it's like an instance of a class, like, you usually dont create smth like `type SomeClassInstance` and use it like `(x: SomeClassInstance) => whatever`, you just use it directly like `(x: SomeClass) => whatever`;
so it's just a convenience / developer ergonomics thing

**Comment by nl0** [comment]
@fiskus 

> There is a control "Undo changes ←" when all files are the same (hashed), but I think, it's ok.

well, it requires some extra logic for comparing trees (maybe overriding final-form's equality check, not sure), and i guess it doesnt worth it right now

> I think that the better strategy for typescript migration is converting small modules to typescript instead of large ones

I agree on that, but I decided to migrate the modules I was modifying and directly importing, bc it was getting hard to reason about that code without proper contracts / typings for state and interfaces

File: catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (fiskus) [reply catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:220]: You can use enum as object key too

```ts
enum State {
  Deleted,
}

entry[State.Deleted]
```

File: catalog/app/containers/Bucket/PackageDialog/PackageDialog.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (fiskus) [reply catalog/app/containers/Bucket/PackageDialog/PackageDialog.tsx]: It will help you replace `any` with real `AsyncResult` type when it's ready: just grep for `FakeAsyncResult` and change it to real one

File: catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:220]: yes, but it wont help with `classes` (or `classes` will have some number keys, which doesnt seem nice). we could use string enums, but they require more boilerplate, so, in the end, i dont see what we gain from that

[the remark this answers is not in the document]
Comment (fiskus) [reply catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx:220]: Profit in letting typescript already know type without manual casting `as const`
